app/routes/questionnaire.py

```diff
@@ -56,17 +61,20 @@
 
 @questionnaire_blueprint.before_request
 @login_required
-@with_questionnaire_store
-def before_questionnaire_request(questionnaire_store):
-    if questionnaire_store.submitted_at:
-        return redirect(url_for("post_submission.get_thank_you"))
-
+def before_questionnaire_request():
     if request.method == "OPTIONS":
         return None
 
     metadata = get_metadata(current_user)
     if not metadata:
-        raise NoQuestionnaireStateException(401)  # pragma: no cover
+        raise NoQuestionnaireStateException(401)
+
+    questionnaire_store = get_questionnaire_store(
+        current_user.user_id, current_user.user_ik
+    )
+
+    if questionnaire_store.submitted_at:
+        return redirect(url_for("post_submission.get_thank_you"))
```

Comment:
We should look to rationalise this in https://trello.com/c/H5vQJuTW/4804-questionnairestate-changes-allow-multiple-submissions, `get_questionnaire_store` should just be made to be consistent with `get_session_store` and not return a store when there is no metadata. Then all you have to do is if check state exists or is submitted, and prevent us from double fetching even if it is cached by flask `g`.